Location chooser using UN boundaries from Fieldmaps.io instead of GeoNames.org

glaroc opened the pull request:
This replaces GeoNames for boundaries from Fieldmaps.io to have better consistency between name selection and extraction of spatial boundaries for each region. The names can be extracted offline. 

This also introduces a new FastAPI at /region/ which, for now, fetches the country and regions lists from the parquet files from Fieldmaps.io and caches them locally. It could be used for other things later. 

An endpoint at /region/geometry/ now allows the direct download of the boundaries in GeoPackage format for any country or region at any level (province, state, county, etc.), if given the proper ISO3 code or region ID, which is the code used internally by Fieldmaps.io.

glaroc committed:
Working first version of python-api with GADM using pygadm package.

glaroc committed:
WIP on the regions form

glaroc committed:
Choosers working

glaroc committed:
GADM functioning in UI - show spinner when loading bbox

glaroc committed:
Update docs for GADM-based selectors

glaroc committed:
OpenAPI specs for /region

jmlord commented:
Correction to the description: the api is under /region/, that routes to a python API.

jmlord commented on python-api/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+COPY --from=builder /app /app
+WORKDIR /app
+
+CMD uvicorn main:app --proxy-headers --port 8001 --host 0.0.0.0 --reload
```
Do a command that starts both titiler and uvicorn.

glaroc replied:
Done

jmlord commented on script-server/api/openapi.yaml:
```diff
@@ -375,7 +375,98 @@ paths:
+                    }
+  /region/regions_list:
+    get:
+      summary: Returns the list of countries from GADM with their ISO3/GID and English names
```
summary is wrong (copy pasted)

glaroc replied:
fixed

jmlord commented on script-server/api/openapi.yaml:
```diff
@@ -375,7 +375,98 @@ paths:
+        "200":
+          description: Geometry of the region in GeoJSON format.
+          content:
+            application/geopackage+sqlite3:
```
geoJSON or GeoPackage? Desc says one and mime type says the other.

jmlord commented on ui/src/components/form/Choosers/CountryRegionMenu.jsx:
```diff
@@ -196,7 +174,7 @@ export default function CountryRegionMenu({
           setSelectedCountry(value);
           setSelectedRegion(null);
-          selectionChanged("country", value?.value ? value.value : null);
+          selectionChanged("country", value?.value ? value : null);
```
ça me semble louche.
voulais-tu dire `value ? value : null`
(mais j'ai pas regardé dans le code la structure de l'objet)

glaroc replied:
C'est ok

jmlord commented on ui/src/components/form/Choosers/utils.jsx:
```diff
@@ -108,21 +104,50 @@ export const paperStyle = (dialog) => {
+  return result;
+};
+
+export const getBBox = async (gid) => {
```
renommer getBBoxAPI, si on veut être constant par rapport aux deux autre méthodes similaires

glaroc replied:
Cette fonction a été retirée

jmlord reviewed (commented):
Code review + test rapide.

Niveau utilisabilité, je remarque que le CRS chooser seul ne me permet pas de chercher Québec Lambert par le nom comme celui qui a les pays le permet.

Petit flash frame si on change de pays et qu'on ouvre à nouveau les régions. C'est bénin, mais on pourrait vider la liste pendant la requête, plutôt que juste la remplacer après.

jmlord committed:
optimize imports

jmlord committed:
Revert "optimize imports"

This reverts commit 036172f05e11cac3d7bf417f762a880ba8c1c8b3.

glaroc committed:
Start migrating to UN vectors

glaroc committed:
Start migrating to UN vectors

glaroc committed:
Rewrite for parquet based UN boundaries

glaroc committed:
Fix API spec

glaroc committed:
Remove bbox api

glaroc committed:
resp instead of unresp

glaroc commented on ui/src/components/form/Choosers/CountryRegionMenu.jsx:
```diff
@@ -36,17 +37,21 @@ export default function CountryRegionMenu({
-    }));
-    setCountryOptions(countryOpts);
+    getCountriesAPI().then((response) => {
+      const unresp = response.data.filter((obj, index, self) =>
```
changed to resp

glaroc commented on script-server/api/openapi.yaml:
```diff
@@ -375,7 +375,98 @@ paths:
+      summary: Returns the list of countries from GADM with their ISO3/GID and English names
+      operationId: getCountriesList
+      responses:
+        "200":
```
Added 404

jmlord commented on python-api/app/countries.json:
I thought this was fetched at runtime? Or it's just missing a proper .gitignore?

jmlord commented on python-api/app/main_api.py:
```diff
@@ -0,0 +1,80 @@
+
+if os.path.exists("/app/countries.json") == False:
+    countries=ddb.sql("SELECT adm0_src, adm0_name, geometry_bbox FROM read_parquet('%s')" % countries_parquet).df()
+    with open('/app/countries.json', 'w') as f:
```
fetched here... (see previous comment)

glaroc replied:
https://github.com/GEO-BON/bon-in-a-box-pipeline-engine/pull/313#issuecomment-3618715714

jmlord commented on http-proxy/conf.d-prod/ngnix.conf:
```diff
@@ -24,7 +24,7 @@ server {
+        proxy_pass http://biab-python-api:8000/;
     }
 
     location /region/ {
```
should this proxy-pass it to http://biab-python-api:8001/region/ [...], so that it's easier to extend the python API for other things than regions in the future?

glaroc replied:
Moved to /region

jmlord commented on python-api/app/main_api.py:
```diff
@@ -0,0 +1,80 @@
+    gs = gpd.GeoSeries.from_wkt(reg["geom"])
+    del reg["geom"]
+    del reg["geometry"]
+    gdf = gpd.GeoDataFrame(reg, geometry=gs, crs="EPSG:4326")
```
toujours dans le même CRS?

glaroc replied:
Je pense que oui

jmlord commented on python-api/app/startup.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+uvicorn main_api:app --proxy-headers --port 8001 --host 0.0.0.0 --reload &
```
ah, ok, la commande est dans le startup. Si c'est une seule commande c'est vrai qu'elle pourrait être mise directement dans le dockerfile, à moins qu'on ait besoin du wrapper bash.

glaroc replied:
Il y a deux commandes. Donc, c'est pour ça le startup.sh

jmlord reviewed (commented):
only reviewed the code changes, yet to be tested.

glaroc commented on compose.yml:
```diff
@@ -70,22 +69,20 @@ services:
-    container_name: biab-tiler
-    image: ghcr.io/developmentseed/titiler:0.18.9
+  biab-python-api:
+    container_name: biab-python-api
```
Added a comment

glaroc commented on python-api/Dockerfile:
```diff
@@ -26,4 +26,11 @@ COPY --from=builder $VIRTUAL_ENV $VIRTUAL_ENV
+ENV WEB_CONCURRENCY=1
+
+CMD sh ./startup.sh
+#CMD uvicorn main:app --proxy-headers --port 8001 --host 0.0.0.0 --reload ; gunicorn -k uvicorn.workers.UvicornWorker ${MODULE_NAME}:${VARIABLE_NAME} --bind 0.0.0.0:${PORT} --workers ${WEB_CONCURRENCY}
```
Removed

glaroc commented on python-api/app/main_api.py:
```diff
@@ -0,0 +1,80 @@
+        if( reg.empty ):
+            raise HTTPException(status_code=404, detail="Region ID not found")
+        fname = reg['adm1_name'].iloc[0].replace(' ','_')
+    print("name:", fname)
```
Removed

glaroc committed:
Updates after code review

glaroc commented:
Ready to merge? @jmlord @JoryGriffith Should be synced with changes in scripts. The region ID is now regionID

glaroc committed:
RegionID

glaroc committed:
Change biab-python-api to python-api

JoryGriffith commented:
Tested this and it is a bit slow to load the packages at first but it works with both the old way of pulling country polygons and the new geoparquet way.

glaroc commented:
> Tested this and it is a bit slow to load the packages at first but it works with both the old way of pulling country polygons and the new geoparquet way.
@JoryGriffith 
What do you mean by the 'old way' ?

JoryGriffith commented:
> > Tested this and it is a bit slow to load the packages at first but it works with both the old way of pulling country polygons and the new geoparquet way.
> > @JoryGriffith
> > What do you mean by the 'old way' ?

With the geoboundaries script that it is in main

glaroc commented on python-api/app/main_api.py:
```diff
@@ -0,0 +1,80 @@
+        if( reg.empty ):
+            raise HTTPException(status_code=404, detail="Country ID not found")
+        fname = reg['adm0_name'].iloc[0].replace(' ','_')
+    if type == 'region':
```
elif

glaroc commented:
Done the last changes. Ready to merge ?

glaroc committed:
Last changes

jmlord commented:
PR comment updated above with the proper API routes. Starting my new (third?) review now

jmlord committed:
updated containers to clean

jmlord commented:
Is it normal that we can choose Quebec Lambert in British Columbia? Looks like something wasn't plugged back.
<img width="411" height="236" alt="image" src="https://github.com/user-attachments/assets/3605bdc8-b59b-4a5c-af43-09b5c216b292" />

Also note the little style problem where the X and the down arrows overlap. It's only for that component.

jmlord commented on python-api/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+WORKDIR /app/
+
+RUN uv pip install fastapi "uvicorn[standard]" requests duckdb pandas geopandas pandas_geojson shapely pygadm fiona
```
Are requests, pandas_geojson and fiona still required? It's not explicitly imported in main_api.py

glaroc replied:
Removed

jmlord committed:
WIP Support for image build
(cannot test until it builds once)

jmlord committed:
python-api is somewhat part of server,
- moved next to server
- added to depends_on

jmlord commented on script-server/api/openapi.yaml:
```diff
@@ -292,8 +287,8 @@ paths:
-        - $ref: '#/components/parameters/stepType'
-        - $ref: '#/components/parameters/runId'
+        - $ref: "#/components/parameters/stepType"
+        - $ref: "#/components/parameters/runId"
```
I'm afraid these formatter changes might conflict with the HPC branch... fingers crossed.

jmlord reviewed (commented):
Review in progress...

The UI works well except for the quirk above, and for the CRS-only chooser, but the latter is out of scope for this PR.

jmlord committed:
Merge remote-tracking branch 'origin/edge' into location_chooser_gadm

jmlord committed:
Regenerated docs

jmlord committed:
string binary, as per openApi spec.
"An API operation can return a file, such as an image or PDF. OpenAPI 3.0 defines file input/output content as type: string with format: binary or format: base64."
+ regenerate client

jmlord committed:
Added example for regions_list

jmlord commented on script-server/api/openapi.yaml:
```diff
@@ -373,9 +368,98 @@ paths:
+            default: country
+        - name: id
+          in: query
+          description: ID of the region to get the geometry for, from the UN regions codes
```
We do not have this id from the regions_list api call.

glaroc replied:
Modified description with adm1_src

jmlord commented on script-server/api/openapi.yaml:
```diff
@@ -373,9 +368,98 @@ paths:
+    get:
+      summary: Returns the geometry of the specified region from GADM in GeoJSON format
+      operationId: getRegionGeometry
+      parameters:
```
Please add working example in both parameters so we can test straight away in the swagger UI.

glaroc replied:
Done

jmlord commented on ui/src/components/form/Choosers/utils.jsx:
```diff
@@ -108,14 +104,28 @@ export const paperStyle = (dialog) => {
 };
 
-export const getStateAPI = async (geonameId) => {
+export const getCountriesAPI = async () => {
```
We have a generated client for this...

glaroc replied:
Updated to generated client

jmlord commented on ui/src/components/form/Choosers/CRSMenu.jsx:
```diff
@@ -40,23 +40,25 @@ export default function CRSMenu({ states, dispatch, value, dialog = false }) {
+          });
+        }
       } else {
         //setCRSList(defaultCRSList);
```
Why is this commented out? 
If we clear the country field, we keep the country suggestions. Is this intended behavior?

glaroc replied:
Removed

jmlord replied:
I thought we had to uncomment it?

jmlord committed:
unused var

jmlord commented on compose.yml:
```diff
@@ -29,6 +28,23 @@ services:
+    container_name: biab-python-api
+    volumes:
+      - ${PIPELINE_REPO_PATH}/output:/output:ro
+      - ${PIPELINE_REPO_PATH}/userdata:/userdata:ro
```
userdata needed?

glaroc replied:
Removed

jmlord reviewed (approved):
Works well! See last comments.

glaroc committed:
Clarify Region ID and return _1 when the ISO has multiple sub-countries

glaroc committed:
Change calls to JS API

glaroc committed:
Move arrow to start

glaroc commented:
> Is it normal that we can choose Quebec Lambert in British Columbia? Looks like something wasn't plugged back. <img alt="image" width="411" height="236" src="https://private-user-images.githubusercontent.com/6223744/534788894-3605bdc8-b59b-4a5c-af43-09b5c216b292.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjgzMjM3NTUsIm5iZiI6MTc2ODMyMzQ1NSwicGF0aCI6Ii82MjIzNzQ0LzUzNDc4ODg5NC0zNjA1YmRjOC1iNTliLTRhNWMtYWY0My0wOWI1YzIxNmIyOTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMTNUMTY1NzM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzcxNmIwZmI0YTlkZWIwZDM1MTYwYWE4OTI1YzI5NDVjZjBmMTdlNDI3NDU0MjIwNzNiZDQwMWUyMjhkNzI3NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Rz8CwA3LOFxY-8pNoItjPpHOTRE_NIRQ97J5KDEnnFE">
> 
> Also note the little style problem where the X and the down arrows overlap. It's only for that component.

Moved it to beginning. It was ok on Firefox, but not on Chrome
